src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -200,41 +234,107 @@ uct_cuda_ipc_open_memhandle_vmm(uct_cuda_ipc_rkey_t *key,
+                                                    CUdeviceptr *mapped_addr)
 {
     CUmemAccessDesc access_desc = {};
-    CUdeviceptr dptr;
     ucs_status_t status;
 
+    *mpool = ucs_malloc(sizeof(CUmemoryPool), "cuda_ipc_mempool");
+    if (*mpool == NULL) {
+        ucs_error("failed to allocate memory for locally mapped mempool");
+        return UCS_ERR_NO_MEMORY;
+    }
+
     status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportFromShareableHandle(
-                &key->ph.pool, (void *)&key->ph.handle.fabric_handle,
+                *mpool, (void *)&key->ph.handle.fabric_handle,
                 CU_MEM_HANDLE_TYPE_FABRIC, 0));
     if (status != UCS_OK) {
-        return status;
+        goto err;
     }
 
-    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(&dptr,
-                key->ph.pool, (CUmemPoolPtrExportData*)&key->ph.ptr));
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(mapped_addr,
+                *(*mpool), (CUmemPoolPtrExportData*)&key->ph.ptr));
     if (status != UCS_OK) {
-        return status;
+        goto err;
     }
 
     status = uct_cuda_ipc_init_access_desc(&access_desc);
     if (status != UCS_OK) {
-        return status;
+        goto err;
     }
 
-    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolSetAccess(key->ph.pool,
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolSetAccess(*(*mpool),
                 &access_desc, 1));
     if (status != UCS_OK) {
-        return status;
+        goto err;
     }
 
-    *mapped_addr = dptr;
-
     return UCS_OK;
+
+err:
+    ucs_free(*mpool);
+    *mpool = NULL;
+    return status;
+}
+
+static ucs_status_t
+uct_cuda_ipc_open_memhandle_mempool(uct_cuda_ipc_rkey_t *key,
+                                    CUdeviceptr *mapped_addr)
+{
+    khash_t(cuda_ipc_rem_mpool_cache) *hash = &uct_cuda_ipc_rem_mpool_cache.hash;
+    CUmemFabricHandle hkey                  = key->ph.handle.fabric_handle;
+    ucs_status_t status                     = UCS_OK;
+    CUmemoryPool *mpool;
+    khiter_t khiter;
+    int khret;
+
+    pthread_rwlock_rdlock(&uct_cuda_ipc_rem_mpool_cache.lock);
+    khiter = kh_get(cuda_ipc_rem_mpool_cache, hash, hkey);
+    if (ucs_likely(khiter != kh_end(hash))) {
+        key->ph.pool = kh_val(hash, khiter);
+        ucs_assert(key->ph.pool != NULL);
+
+        status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(mapped_addr,
+                    key->ph.pool, (CUmemPoolPtrExportData*)&key->ph.ptr));
+        goto err;
+    }
+
+    /* acquire writer lock as we did not find cached mpool in hashmap */
+    pthread_rwlock_unlock(&uct_cuda_ipc_rem_mpool_cache.lock);
+    pthread_rwlock_wrlock(&uct_cuda_ipc_rem_mpool_cache.lock);
+
+    khiter = kh_put(cuda_ipc_rem_mpool_cache, hash, hkey, &khret);
+    if ((khret == UCS_KH_PUT_BUCKET_EMPTY) ||
+        (khret == UCS_KH_PUT_BUCKET_CLEAR)) {
+        status = cuda_ipc_rem_mpool_cache_create(key, &mpool, mapped_addr);
```

Comment:
seems like you can avoid using pointer to pointer and pass just a pointer to some local `CUmemoryPool` var, which will be stored in hash anyway

Reply:
right, let me fix that, since I added this function